### PR TITLE
WIP: add audit logs

### PR DESCRIPTION
**Pull request by Nautigsam**
This is a WIP to implement audit logs. This is an attempt to make a first step towards a log panel inside admin tools (see #826 for original request).
I already added audit logs for videos and comments but I wanted to get some review of the implementation before generalizing it to other entities.

As an example, here is the log I get when I add a video:
```json
{"level":"audit","message":"{\"user\":1,\"domain\":\"videos\",\"action\":\"create\",\"video-tags\":[],\"video-uuid\":\"6edd2620-3c68-4e43-9cd5-f7b3a1797d03\",\"video-id\":3,\"video-createdAt\":\"2018-07-23T11:15:49.244Z\",\"video-updated
At\":\"2018-07-23T11:15:49.244Z\",\"video-publishedAt\":\"2018-07-23T11:15:48.203Z\",\"video-description\":null,\"video-duration\":287,\"video-isLocal\":true,\"video-name\":\"SketchFinal\",\"video-thumbnailPath\":\"/static/thumbnails/6ed
d2620-3c68-4e43-9cd5-f7b3a1797d03.jpg\",\"video-previewPath\":\"/static/previews/6edd2620-3c68-4e43-9cd5-f7b3a1797d03.jpg\",\"video-nsfw\":false,\"video-waitTranscoding\":true,\"video-account-id\":2,\"video-account-uuid\":\"f357b1f7-d34a
-4940-9966-d79d1ed9ad50\",\"video-account-name\":\"root\",\"video-channel-id\":1,\"video-channel-uuid\":\"e711fb0d-7dfa-49c2-8fb1-5599927fd672\",\"video-channel-name\":\"e711fb0d-7dfa-49c2-8fb1-5599927fd672\",\"video-support\":null,\"vid
eo-commentsEnabled\":true}","timestamp":"2018-07-23T11:15:49.286Z","label":"localhost:9000"}
```

Yes, that's pretty big. I can actually make an even more strict filter if you think some fields are useless. I also had a hard time deciding how I could identify the user so I user the userId from the OAuth token but maybe there is a better way?

### CODE REVIEW

File: server/controllers/api/videos/index.ts [hunk 1]
```diff
@@ -298,6 +302,7 @@ async function updateVideo (req: express.Request, res: express.Response) {
   }
 
   try {
+    let videoInstanceUpdated
```
Comment (Chocobozzz) [inline server/controllers/api/videos/index.ts]: Why?

**Comment by Chocobozzz** [comment]
I think we would want to audit remote objects too (coming from the federation) (see `/server/lib/activitypub/process` directory).

>  I can actually make an even more strict filter if you think some fields are useless

``` 
const videoKeysToKeep = [
  'tags',
  'category',
  'licence',
  'language',
  'privacy',
  'description',
  'name',
  'nsfw',
  'account-uuid',
  'account-name',
  'channel-uuid',
  'channel-name',
  'support'
]

const commentKeysToKeep = [
  'text',
  'video-name',
  'account-uuid',
  'account-name'
]
```

Should be enough for now.

> I also had a hard time deciding how I could identify the user so I user the userId from the OAuth token but maybe there is a better way?

Prefer using the account username + hostname (if remote): `res.locals.oauth.token.User.Account.Actor.preferredUsername + '@' + res.locals.oauth.token.User.Account.Actor.getHost()` (please create a method for this below `getWebfingerUrl` in actor model)

File: server/controllers/api/videos/index.ts (hunk 1 above)
Comment (Nautigsam) [reply server/controllers/api/videos/index.ts]: I don't have strong opinion on this change. I'm sure there is a better way. I could simply make the log inside the transaction.

Comment (Chocobozzz) [reply server/controllers/api/videos/index.ts]: Oh, did not see the log was outside! But yes put it just below `federateVideoIfNeeded` (+ logger.info too please)

**Comment by Nautigsam** [comment]
I pushed some corrections according to your remarks. I will now use it in as many entities as I can. 

> I think we would want to audit remote objects too

I tend to disagree with you here. I think the use case is more like "As an admin, I use audit logs to monitor writing actions of users connected on my instance". I'm afraid that logging every actions from remote instances will add a lot of noise for actions the admin can not do anything about.

**Comment by Chocobozzz** [comment]
> I tend to disagree with you here. I think the use case is more like "As an admin, I use audit logs to monitor writing actions of users connected on my instance". I'm afraid that logging every actions from remote instances will add a lot of noise for actions the admin can not do anything about.

I understand but remote objects are as important as you local objects since they are displayed on your instance, with almost no differences with your local objects. But let's ping @firePowi

**Comment by Chocobozzz** [comment]
Very nice work @Nautigsam! Is it still a WIP?

**Comment by Nautigsam** [comment]
Some logs are missing because of a bug I had. I needed some refactoring in another place. If the core module and the current uses seem ok, then you can merge it I guess.

**Comment by Chocobozzz** [comment]
:+1: Thanks @Nautigsam. Let's merge it now to avoid conflicts.

> Some logs are missing because of a bug I had.

What is the bug and what logs are missing?

**Comment by Nautigsam** [comment]
> What is the bug and what logs are missing?
I got a bug when I added logs for videos blacklists. I will try again but it's not trivial to test as it needs a second instance running (#825 could be handy in that case).

Captions logs are missing. I also wanted to log when an admin add instances to follow, but I don't yet fully understand that part and maybe it would require some refactoring.

**Comment by FirePowi** [comment]
Sorry for late respond (wasn't around)

So I might more agree with @Nautigsam → Local log.

If someone from another instance leave a comment on the local instance → log it
If someone from another instance publish a video on that another instance → do not log it / leave the choice to admin to log it or not [options takes time so… Maybe first one]
If someone from my instance leave a comment on another instance → log it

That's what I wanted but don't know if it's what everyone want. So I just give my point of view, give yours, let's discuss ?

**Comment by Chocobozzz** [comment]
We could add an enum field inside the object (`local`, 'to_remote`, `from_remote`); it would cover all the needs if we give the ability on the client side to filter the entries.

**Pull request merged**